File: BEmuJava/bemu/src/main/java/com/bloomberglp/blpapi/HistoricElementFieldDataArray.java
//------------------------------------------------------------------------------
// <copyright project="BEmu_maven" file="/BEmu_maven/bemu/src/main/java/com/bloomberglp/blpapi/HistoricElementFieldDataArray.java" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

package com.bloomberglp.blpapi;

import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import com.bloomberglp.blpapi.Datetime;
import com.bloomberglp.blpapi.Name;
import com.bloomberglp.blpapi.Element;
import com.bloomberglp.blpapi.SchemaTypeDefinition;

public class HistoricElementFieldDataArray extends Element
{
	private final List<HistoricElementFieldData> _fieldData;
	
	HistoricElementFieldDataArray(Map<Datetime, Map<String, Object>> fieldData) throws Exception
	{
		this._fieldData = new ArrayList<HistoricElementFieldData>();
		
		for (Map.Entry<Datetime, Map<String, Object>> item : fieldData.entrySet())
		{
			HistoricElementFieldData elmFieldData = new HistoricElementFieldData(item.getKey(), item.getValue());
			this._fieldData.add(elmFieldData);
		}
	}
    
    public SchemaTypeDefinition typeDefinition() throws Exception
    {
    	return new SchemaTypeDefinition(this.datatype(), new Name("HistoricalDataRow"));
    }
	
	public Element getValueAsElement(int index)
	{
		return this._fieldData.get(index);
	}
	
    public Name name() throws Exception
    {
    	return new Name("fieldData");
    }
    
    public int numValues()
    {
    	return this._fieldData.size();
    }
    
    public int numElements()
    {
    	return 0;
    }
    
    public boolean isComplexType()
    {
    	return true;
    }
    
    public boolean isArray()
    {
    	return true;
    }
    
    public boolean isNull()
    {
    	return false;
    }
    
    protected StringBuilder prettyPrint(int tabIndent) throws Exception
    {
        String tabs = com.bloomberglp.blpapi.IndentType.Indent(tabIndent);
        StringBuilder result = new StringBuilder();

        result.append(String.format("%s%s[] = {%s", tabs, this.name().toString(), System.getProperty("line.separator")));
        
        for(int i = 0; i < this._fieldData.size(); i++)
        {
        	result.append(this._fieldData.get(i).prettyPrint(tabIndent + 1));
        }

        result.append(String.format("%s}%s", tabs, System.getProperty("line.separator")));
        return result;
    }
}

File: BEmu/HistoricalDataRequest/HistoricElementFieldDataArray.cs
//------------------------------------------------------------------------------
// <copyright project="BEmu_csh" file="HistoricalDataRequest/HistoricElementFieldDataArray.cs" company="Jordan Robinson">
//     Copyright (c) 2013 Jordan Robinson. All rights reserved.
//
//     The use of this software is governed by the Microsoft Public License
//     which is included with this distribution.
// </copyright>
//------------------------------------------------------------------------------

namespace Bloomberglp.Blpapi.HistoricalDataRequest
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;

    internal class HistoricElementFieldDataArray : Element
    {
        private readonly List<HistoricElementFieldData> _fieldData;

        internal HistoricElementFieldDataArray(Dictionary<DateTime, Dictionary<string, object>> fieldData)
        {
            this._fieldData = new List<HistoricElementFieldData>();

            foreach (var dtDate in fieldData)
	        {
                HistoricElementFieldData elmFieldData = new HistoricElementFieldData(dtDate.Key, dtDate.Value);
                this._fieldData.Add(elmFieldData);
	        }
        }

        public override Element GetValueAsElement(int index) { return this._fieldData[index]; }
        public override object this[int index] { get { return this._fieldData[index]; } }

        public override Name Name { get { return new Name("fieldData"); } }
        public override int NumElements { get { return 0; } }
        public override int NumValues { get { return this._fieldData.Count; } }
        public override bool IsComplexType { get { return false; } }
        public override bool IsArray { get { return true; } }

        public override SchemaTypeDefinition TypeDefinition { get { return new SchemaTypeDefinition(this.Datatype, new Name("HistoricalDataRow")); } }

        internal override StringBuilder PrettyPrint(int tabIndent, bool surroundValueWithQuotes = false)
        {
            string tabs = Types.IndentType.Indent(tabIndent);
            StringBuilder result = new StringBuilder();

            result.AppendFormat("{0}{1}[] = {{{2}", tabs, this.Name, Environment.NewLine);
            foreach (var item in this._fieldData)
            {
                result.Append(item.PrettyPrint(tabIndent + 1));
            }
            result.AppendFormat("{0}}}{1}", tabs, Environment.NewLine);

            return result;
        }
    }
}
